flink-tests/src/test/java/org/apache/flink/test/checkpointing/ZooKeeperHighAvailabilityITCase.java

```diff
@@ -0,0 +1,387 @@
+		final JobID jobID = Preconditions.checkNotNull(jobGraph.getJobID());
+
+		// Retrieve the job manager
+		final ActorGateway jobManager = Await.result(cluster.leaderGateway().future(), deadline.timeLeft());
+
+		cluster.submitJobDetached(jobGraph);
+
+		// wait until we did some checkpoints
+		waitForCheckpointLatch.await();
+
+		// mess with the HA directory so that the job cannot restore
+		File movedCheckpointLocation = temporaryFolder.newFolder();
+		int numCheckpoints = 0;
+		File[] files = haStorageDir.listFiles();
+		assertNotNull(files);
+		for (File file : files) {
+			if (file.getName().startsWith("completedCheckpoint")) {
+				assertTrue(file.renameTo(new File(movedCheckpointLocation, file.getName())));
+				numCheckpoints++;
+			}
+		}
+		assertTrue(numCheckpoints > 0);
+
+		failInCheckpointLatch.trigger();
+
+		// Ensure that we see at least one cycle where the job tries to restart and fails.
+		Future<JobStatus> jobStatusFuture = FutureUtils.retrySuccessful(
+			new Callable<Future<JobStatus>>() {
+				@Override
+				public Future<JobStatus> call(){
+					return getJobStatus(jobManager, jobID, TEST_TIMEOUT);
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus){
+					return jobStatus == JobStatus.RESTARTING;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.RESTARTING, jobStatusFuture.get());
+
+		jobStatusFuture = FutureUtils.retrySuccessful(
+			new Callable<Future<JobStatus>>() {
+				@Override
+				public Future<JobStatus> call() {
+					return getJobStatus(jobManager, jobID, TEST_TIMEOUT);
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus) {
+					return jobStatus == JobStatus.FAILING;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.FAILING, jobStatusFuture.get());
+
+		// move back the HA directory so that the job can restore
+		CheckpointBlockingFunction.afterMessWithZooKeeper.set(true);
+
+		files = movedCheckpointLocation.listFiles();
+		assertNotNull(files);
+		for (File file : files) {
+			if (file.getName().startsWith("completedCheckpoint")) {
+				assertTrue(file.renameTo(new File(haStorageDir, file.getName())));
+			}
+		}
+
+		// now the job should be able to go to RUNNING again and then eventually to FINISHED
+		jobStatusFuture = FutureUtils.retrySuccessful(
+			new Callable<Future<JobStatus>>() {
+				@Override
+				public Future<JobStatus> call() {
+					return getJobStatus(jobManager, jobID, TEST_TIMEOUT);
+				}
+			},
+			new FilterFunction<JobStatus>() {
+				@Override
+				public boolean filter(JobStatus jobStatus) {
+					return jobStatus == JobStatus.FINISHED;
+				}
+			},
+			deadline,
+			TestingUtils.defaultExecutor());
+		assertEquals(JobStatus.FINISHED, jobStatusFuture.get());
+
+		// make sure we saw a successful restore
+		successfulRestoreLatch.await();
```

Comment:
If you are in state `FINISHED` here, you must have successfully restored. If you did not successfully restore, this latch will block forever and the test will time out.

Reply:
Yes, I'll remove that.